playground/css-sourcemap/__tests__/css-sourcemap.spec.ts

```diff
@@ -69,6 +69,7 @@ describe.runIf(isServe)('serve', () => {
     expect(formatSourcemapForSnapshot(map)).toMatchInlineSnapshot(`
       {
         "mappings": "AAAA;EACE,UAAU;AACZ;;ACAA;EACE,UAAU;AACZ",
+        "sourceRoot": "/root",
```

Comment:
I don't think this is correct? The file is absolute, so it might work for the browser (as a pathname), but this will be incorrect in Node.js (should be a filepath).

Why do we need to specify `sourceRoot` anyway?

Reply:
When the `sources` are relative paths, provide a `sourceRoot` to assemble the absolute paths. I think this is useful for user to locate the error source.

The throw error before PR, I think `/src/error.js:2:9` is not easy to find the source file especially when using relative path.
```
# sourcemaps enabled: true
Error: e
    at Module.error (/src/error.js:2:9)
    at file:///Users/fi3ework/Downloads/vite-main/playground/ssr-html/test-stacktrace.js:36:7
```

After PR:
All types of imports will using an absolute path in error stack.

```
# sourcemaps enabled: true
# source file extension: js
Error: e
    at Module.error (/Users/fi3ework/OSS/vite/playground/ssr-html/src/error.js:2:9)
    at file:///Users/fi3ework/OSS/vite/playground/ssr-html/test-stacktrace.js:44:9
Error: e
    at Module.error (/Users/fi3ework/OSS/vite/playground/ssr-html/src/error.js:2:9)
    at file:///Users/fi3ework/OSS/vite/playground/ssr-html/test-stacktrace.js:44:9
Error: e
    at Module.error (/Users/fi3ework/OSS/vite/playground/ssr-html/src/error.js:2:9)
    at file:///Users/fi3ework/OSS/vite/playground/ssr-html/test-stacktrace.js:44:9
```

Reply:
Source route doesn't have to be absolute as far as I know.

In your example code will not work correctly in SSR because it's not absolute to the file system.

How does it know that `/root` is actually relative to the root of the project and not an absolute path on the file system?

Reply:
> Source route doesn't have to be absolute as far as I know.

Yeah, it doesn't have to be. But it will be easy to locate the source file IMO 🤔.

> In your example code will not work correctly in SSR because it's not absolute to the file system.

The example generated by running `pnpm test-serve -t stacktrace` in ` playground/ssr-html`. Could you explain in more detail, Thanks! 

> How does it know that /root is actually relative to the root of the project and not an absolute path on the file system?

`/root` is replaced by an absolute path in E2E test to normalize the actual paths in different runners. If not matched, it should keep as-is.

Reply:
@fi3ework Would changing `sourceURL=${mod.url}` to `sourceURL=${mod.file}` work?
I guess Node.js is treating the original file path as `/src/error.js`. We want Node.js to treat the original file path as `/Users/fi3ework/OSS/vite/playground/ssr-html/src/error.js` (or `file:///Users/fi3ework/OSS/vite/playground/ssr-html/src/error.js`).
https://github.com/vitejs/vite/blob/a8c7eb24a2b7d8dc765d59758e2c6b930eb90af1/packages/vite/src/node/ssr/ssrModuleLoader.ts#L225

Reply:
@sapphi-red Basically could work except for Windows 😅. Check out  https://github.com/vitejs/vite/actions/runs/4939126748/jobs/8829577947?pr=12902#step:14:50. The error location was incorrectly resolved to 4 instead of 2 for two cases (others are good). I don't think I can fix it quickly without a Windows device. 😅 (had reverted the tweak)

Reply:
According to the sourcemap spec, when the `sources` are relative paths, it will be resolved from the place of the file. So it should work without specifing `sourceRoot`. Also setting `sourceRoot` will cause a problem that #12079 solved.

Vite was lacking support for resolving relative `sources` and that was causing the stacktrace to be like that. I pushed some commits implementing that. Coincidentally I found the reason of https://github.com/vitejs/vite/pull/12902#discussion_r1170477865 and fixed that too. It would be nice if you could review my change.

Reply:
Using `id` is a much more clever solution than `url`. Nice work! LGTM.